File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Kyle's solution: HashTable, Heap
        # Time Complexity: O(N logK)
        # Space Complexity: O(K)
        # N: size of nums array

        prioryQ = list() # Min heap

        for num, freq in Counter(nums).items():
            heapq.heappush(prioryQ, (freq, num))
            if len(prioryQ) > k:
                heapq.heappop(prioryQ)

        return [num for _, num in prioryQ]

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        /*
            Algorithms & DS: #Hashing
            Time: O(N), Space: O(N)
         */

        HashMap<Integer, Integer> count = new HashMap<>();
        HashMap<Integer, Integer> rank = new HashMap<>();
        int[] ans = new int[k];

        for(int num : nums){
            if(!rank.containsKey(num)){
                int initRank = rank.size();
                rank.put(num, initRank);

                if(initRank < k){
                    ans[initRank] = num;
                }
            }

            count.put(num, count.getOrDefault(num, 0) + 1);

            int currRank = rank.get(num);
            int currCount = count.get(num);
            int betterRank = Math.min(k - 1, currRank - 1);

            while (betterRank > -1 && count.get(ans[betterRank]) < currCount){
                int nextNum = ans[betterRank];
                rank.put(nextNum, currRank);
                rank.put(num, betterRank);

                // swap rank
                ans[betterRank] = num;

                if(currRank < k){
                    ans[currRank] = nextNum;
                }

                currRank = betterRank;
                betterRank = currRank - 1;
            }
        }
        
        return ans;
    }
}